### ISSUE TITLE
How to push history state when using htmx.ajax()?

### ISSUE THREAD

**Issue by lachdoug** [bigskysoftware/htmx#2456]
I'd like to update browser history after a `htmx.ajax()` call. 

`htmx.ajax()` does not have an option to update history state automatically, like we have when using the `hx-push-url` attribute.

I would have expected this to work.

```javascript
htmx.ajax("GET", url, {
    target: "#content",
    select: "#content",
    swap: "outerHTML",
}).then(() => {
    history.pushState({ htmx: true }, "", url);
});
```

However, when back button is pressed after updating content, the browser location updates, but the previous content is not displayed.

Is there an idiomatic way to integrate `htmx.ajax()` and ` history.pushState`?

**Comment by lachdoug** [comment]
Is that function exposed in the public API? I can see that function in the source code, but can't see how to access it from my application js.

**Comment by p-baum** [comment]
Set the `source`  attribute to the element where you want to read `hx-push-url` (or any other setting) from:
```
htmx.ajax("GET", url, {
    target: "#content",
    select: "#content",
    swap: "outerHTML",
    source: htmx.find("#my-pretend-source-elem")
})
```

**Comment by MichaelWest22** [comment]
Added a possible PR for this feature

**Comment by pabmann** [comment]
Have you tried using the source attribute available in ajax and having hx-push-url set on the source element?
I seem to remember this working for me.  Or perhaps it was another hx- attribute.  I forget now

**Comment by MichaelWest22** [comment]
Yes this is a good workaround to the problem currently i've suggested many times.  It will use the source dummy elements attributes in the ajax request but it makes it harder to use just the ajax api alone as you also have to create dummy real elements with the custom attributes you need or find one you can reuse as the source. So i'm trying to see if we can instead add this feature to the ajax api itself because it is actually needed internally by hx-location to fix a edge case bug anyway.

**Comment by 1cg** [comment]
@MichaelWest22 this is fixed in four (as you know) you think that that PR is worth integrating into two just to have?  Maybe align the option names?

**Comment by 1cg** [comment]
dupe of https://github.com/bigskysoftware/htmx/issues/2744
